subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java

```diff
@@ -119,41 +128,105 @@ protected Class<?> resolveType(Object value) {
-                handler.accept("$" + counter++, o);
+                String prefix = o instanceof Named ? ((Named) o).getName() : "";
+                String name = prefix + "$" + counter++;
+                checkNotNullNestedCollectionValue(qualifiedName, name, o);
+                handler.accept(name, o);
+            }
+        }
+
+        private void checkNotNullNestedCollectionValue(@Nullable String parentQualifiedName, String name, @Nullable Object value) {
+            if (value == null) {
+                throw new IllegalStateException(getNullNestedCollectionValueExceptionMessage(parentQualifiedName, name));
             }
         }
 
+        private String getNullNestedCollectionValueExceptionMessage(@Nullable String parentQualifiedName, String name) {
+            return String.format("Null or absent is not allowed for the nested property '%s', since it's an element of a collection", getQualifiedName(parentQualifiedName, name));
+        }
+
+        @Override
+        protected Optional<Object> getNestedChild(Object parent, String childQualifiedName, PropertyMetadata propertyMetadata, InstanceMetadataVisitor visitor) {
+            return tryUnpackNested(
+                childQualifiedName,
+                visitor,
+                () -> visitMissingNested(visitor, childQualifiedName, propertyMetadata),
+                () -> getChild(parent, propertyMetadata)
+            );
+        }
+
+        private void visitMissingNested(InstanceMetadataVisitor visitor, String qualifiedName, PropertyMetadata propertyMetadata) {
+            visitor.visitMissingNested(qualifiedName, propertyMetadata);
+        }
+
         @Override
         protected Optional<Object> getChild(Object parent, PropertyMetadata property) {
+            Method method = property.getGetterMethod();
             try {
-                return Optional.ofNullable(property.getGetterMethod().invoke(parent));
+                // TODO: Move method.setAccessible(true) to PropertyMetadata
+                method.setAccessible(true);
+                return Optional.ofNullable(method.invoke(parent));
+            } catch (InvocationTargetException e) {
+                throw UncheckedException.throwAsUncheckedException(e.getCause());
             } catch (Exception e) {
-                // TODO Handle this
-                throw new RuntimeException(e);
+                throw new GradleException(String.format("Could not call %s.%s() on %s", method.getDeclaringClass().getSimpleName(), method.getName(), parent), e);
             }
         }
+
+        private Optional<Object> tryUnpackNested(String childQualifiedName, InstanceMetadataVisitor visitor, Runnable onMissingValue, Supplier<Optional<Object>> unpacker) {
+            Optional<Object> value;
+            try {
+                value = unpacker.get();
+            } catch (Exception e) {
+                visitor.visitUnpackNestedError(childQualifiedName, e);
+                return Optional.empty();
```

Comment:
It would be nicer, but I think it would not be ok, since whoever implements `visitUnpackNestedError` would have to know the implementation details of `AbstractTypeMetadataWalker`